zh-cn: init translation of svelte deployment (#18302)
mdn/translated-content: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_deployment_next/index.md

   ✅  Production: https://mdn-svelte-tutorial.vercel.app [copied to clipboard] [19s]
      Deployed to production. Run `vercel --prod` to overwrite later (https://vercel.link/2F).
      To change the domain or build command, go to https://zeit.co/opensas/mdn-svelte-tutorial/settings
   ```

4. 接受所有默认设置，这没有任何问题。
5. 当部署完成后，在浏览器中打开“生产环境” URL，你会看到已经部署的应用！

你还可以从 [GitHub](https://github.com/)、[GitLab](https://about.gitlab.com/) 或 [BitBucket](https://bitbucket.org/product) [将 Svelte git 项目导入](https://vercel.com/import/svelte)到 Vercel 中。

> **备注：** 你可以通过 `npm i -g vercel` 全局安装 Vercel，这样你就不必使用 `npx` 来运行它了。

### 自动部署到 GitLab Pages

关于托管静态文件，有几个在线服务可以让你在每次推送代码到 git 仓库时，自动部署你的网站。大多数服务都涉及设置一个部署流程，该流程在每次 `git push` 时触发，负责构建和部署你的网站。

我们将把我们的待办事项应用部署到 [GitLab Pages](https://about.gitlab.com/stages-devops-lifecycle/pages/)，借此来演示这一点。

1. 首先，你需要[在 GitLab 注册](https://gitlab.com/users/sign_up)，然后[创建一个新项目](https://gitlab.com/projects/new)。给你的新项目取一个简短、简单的名称，比如“mdn-svelte-todo”。你将获得指向你的新 GitLab 存储库的远程 URL，例如 `git@gitlab.com:[你的用户名]/[你的项目名].git`。
2. 在开始上传内容到你的 git 存储库之前，最好的做法是添加 `.gitignore` 文件，告诉 git 哪些文件应该排除在源代码控制之外。在我们的例子中，我们将在你本地项目的根文件夹中创建 `.gitignore` 文件，内容如下：

   ```bash
   node_modules/
   ```

> 3. 现在让我们回到 GitLab。在创建新仓库后，GitLab 将向你打招呼，并解释了上传现有文件的不同选项。按照 _Push an existing folder_ （推送现有文件夹）标题下的步骤进行操作：

Author: ```suggestion
3. 现在让我们回到 GitLab。在创建新仓库后，GitLab 将向你打招呼，并解释了上传现有文件的不同选项。按照 *Push an existing folder*（推送现有文件夹）标题下的步骤进行操作：
```

   ```bash
   cd your_root_directory # 进入项目的根文件夹
   git init
   git remote add origin https://gitlab.com/[你的用户名]/mdn-svelte-todo.git
   git add .
   git commit -m "Initial commit"
   git push -u origin main
   ```

>    > **备注：** 你可以使用 [git 协议](https://git-scm.com/book/zh/v2/%E6%9C%8D%E5%8A%A1%E5%99%A8%E4%B8%8A%E7%9A%84Git-%E5%8D%8F%E8%AE%AE#_the_git_protocol)替代 `https`，它更快速，而且无需每次访问远程仓库时输入用户名和密码。要使用 git 协议，你需要[创建一个 SSH 密钥对](https://docs.gitlab.com/ee/user/ssh.html#generate-an-ssh-key-pair)。你的远程仓库 URL 将是这样的：`git@gitlab.com:[你的用户名]/mdn-svelte-todo.git`。

Reviewer: ```suggestion
   > **备注：** 你可以使用 [git 协议](https://git-scm.com/book/zh/v2/服务器上的-Git-协议#_git_协议)替代 `https`，它更快速，而且无需每次访问远程仓库时输入用户名和密码。要使用 git 协议，你需要[创建一个 SSH 密钥对](https://docs.gitlab.cn/jh/user/ssh.html#生成-ssh-密钥对)。你的远程仓库 URL 将是这样的：`git@gitlab.com:[你的用户名]/mdn-svelte-todo.git`。
```

按照以上步骤初始化本地 git 仓库，然后将远程仓库设置为 GitLab 上的仓库，接下来将所有文件提交到本地 git 仓库，然后将它们推送到 GitLab 上的远程仓库。

GitLab 使用名为 GitLab CI/CD 的内置工具来构建你的网站并将其发布到 GitLab Pages 服务器。GitLab CI/CD 运行的脚本序列用于完成这个任务，这些脚本是在名为 `.gitlab-ci.yml` 的文件中创建和修改的。配置文件中的一个特定作业叫做 `pages`，它让 GitLab 知道你正在部署一个 GitLab Pages 网站。

现在我们来尝试一下。

1. 在项目的根目录下创建 `.gitlab-ci.yml` 文件，并将以下内容复制进去：

   ```yaml
   image: node:latest
   pages:
     stage: deploy
     script:
       - npm install
       - npm run build
     artifacts:
       paths:
         - public
     only:
       - main
   ```

> 在这里，我们告诉 GitLab 使用最新版本的 Node 镜像来构建我们的应用程序。接下来，我们声明了一个 `pages` 作业，以启用 GitLab Pages。每当我们对仓库进行推送时，GitLab 将运行 `npm install` 和 `npm run build` 来构建我们的应用程序。我们还告诉 GitLab 部署 `public` 文件夹的内容。在最后一行，我们配置 GitLab 仅在对主分支进行推送时重新部署我们的应用程序。

Author: ```suggestion
   在这里，我们告诉 GitLab 使用最新版本的 Node 镜像来构建我们的应用程序。接下来，我们声明了一个 `pages` 作业，以启用 GitLab Pages。每当我们对仓库进行推送时，GitLab 将运行 `npm install` 和 `npm run build` 来构建我们的应用程序。我们还告诉 GitLab 部署 `public` 文件夹的内容。在最后一行，我们配置 GitLab 仅在对主分支进行推送时重新部署我们的应用程序。
```

> 2. 由于我们的应用程序将在子目录发布（例如 `https://your-user.gitlab.io/mdn-svelte-todo`），我们需要使 `public/index.html` 文件中对 JavaScript 和 CSS 文件的引用变为相对路径。为了做到这一点，我们只需从 `/global.css`、`/build/bundle.css` 和 `/build/bundle.js` 的 URL 中去掉前面的斜杠(`/`)，像这样：

Author: ```suggestion
2. 由于我们的应用程序将在子目录发布（例如 `https://your-user.gitlab.io/mdn-svelte-todo`），我们需要使 `public/index.html` 文件中对 JavaScript 和 CSS 文件的引用变为相对路径。为了做到这一点，我们只需从 `/global.css`、`/build/bundle.css` 和 `/build/bundle.js` 的 URL 中去掉前面的斜杠（`/`），像这样：
```

   <script defer src="build/bundle.js"></script>
   ```

   现在进行这个操作。

3. 现在，我们只需要提交并推送我们的更改到 GitLab。通过运行以下命令来完成：

   ```bash
   git add public/index.html
   git add .gitlab-ci.yml
   git commit -m "Added .gitlab-ci.yml file and fixed index.html absolute paths"
   git push
   ```

每当有作业在运行时，GitLab 将显示一个图标，显示作业的进程。单击它以检查作业的输出。

![gitlab 截图显示部署的提交，该提交添加了 gitlab ci 文件，并更改了 bundle 路径为相对路径](01-gitlab-pages-deploy.png)

你还可以从 GitLab 项目的 _CI / CD_ > _Jobs_ 菜单选项中，检查当前和之前作业的进度。

![在 gitlab 界面中显示的 gitlab ci 作业，运行了许多命令](02-gitlab-pages-job.png)

一旦 GitLab 完成构建和发布你的应用程序，该应用程序将可以通过 `https://your-user.gitlab.io/mdn-svelte-todo/` 访问；在我的例子中是 `https://opensas.gitlab.io/mdn-svelte-todo/`。你可以在 GitLab 的用户界面中检查页面的 URL，具体操作请参阅 _Settings_ > _Pages_ 菜单选项。

通过这个配置，每当你推送更改到 GitLab 仓库，应用程序将自动重建并部署到 GitLab Pages。

## 进一步了解 Svelte

在这个部分，我们将为你提供一些资源和项目，以进一步学习 Svelte。

### Svelte 文档

要深入了解 Svelte 并学习更多关于它的知识，你一定要访问 [Svelte 主页](https://svelte.dev/)。在那里，有[许多文章](https://svelte.dev/blog)解释 Svelte 的理念。如果你还没有这样做，一定要完成 [Svelte 交互式教程](https://learn.svelte.dev/)。我们已经涵盖了大部分内容，所以你不需要花费太多时间来完成它，你可以将其视为实践！

你还可以查阅 [Svelte API 文档](https://svelte.dev/docs)和可用的[示例代码](https://svelte.dev/examples/hello-world)。

> 要了解 Svelte 作者创建它背后的动力，你应该观看 [Rich Harris](https://twitter.com/Rich_Harris) 在 YouTube 上的 [Rethinking reactivity](https://www.youtube.com/watch?v=AdNJ3fydeao&t=47s)（重新思考响应式）演讲。他是 Svelte 的创作者，所以他对此有很多见解。你还可以在这里找到交互式幻灯片，这些幻灯片显然是使用 Svelte 构建的。如果你喜欢这个演讲，你还可以观看 Rich Harris 在 [JSCAMP 2019](https://jscamp.tech/2019/) 上的演讲 [The Return of 'Write Less, Do More'](https://www.youtube.com/watch?v=BzX4aTRPzno)（“写的少，做的多”的回归）。

Reviewer: ```suggestion
要了解 Svelte 作者创建它背后的动机，你应该观看 [Rich Harris](https://twitter.com/Rich_Harris) 在 YouTube 上的 [Rethinking reactivity（重新思考响应式）](https://www.youtube.com/watch?v=AdNJ3fydeao&t=47s)演讲。他是 Svelte 的创作者，所以他对此有很多见解。你还可以在这里找到交互式幻灯片，当然，这些幻灯片是使用 Svelte 构建的。如果你喜欢这个演讲，你还可以观看 Rich Harris 在 [JSCAMP 2019](https://jscamp.tech/2019/) 上的演讲 [The Return of 'Write Less, Do More'（“写的少，做的多”的回归）](https://www.youtube.com/watch?v=BzX4aTRPzno)。
```
